File: src/main/scala/com/cloudera/sparkts/ARIMA.scala
/**
 * Copyright (c) 2015, Cloudera, Inc. All Rights Reserved.
 *
 * Cloudera, Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"). You may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for
 * the specific language governing permissions and limitations under the
 * License.
 */

package com.cloudera.sparkts

import breeze.linalg._

import org.apache.commons.math3.analysis.MultivariateFunction
import org.apache.commons.math3.optim.{SimpleBounds, MaxEval, MaxIter, InitialGuess}
import org.apache.commons.math3.optim.nonlinear.scalar.{GoalType, ObjectiveFunction}
import org.apache.commons.math3.optim.nonlinear.scalar.noderiv.BOBYQAOptimizer
import org.apache.commons.math3.random.RandomGenerator
import org.apache.commons.math3.stat.regression.OLSMultipleLinearRegression


object ARIMA {
  /**
   * Given a time series, fit a non-seasonal ARIMA model of order (p, d, q), where p represents
   * the autoregression terms, d represents the order of differencing, and q moving average error
   * terms. If includeIntercept is true, the model is fitted with a mean. The user can specify
   * the method used to fit the parameters. The current implementation only supports "css",
   * meaning conditional sum of squares. In order to select the appropriate order of the model,
   * users are advised to visualize inspect ACF and PACF plots. Finally, current implementation
   * does not verify if the parameters fitted are stationary/invertible. Given this, it is suggested
   * that users inspect the resulting model. However, if parameters are not sane, it is likely
   * that the order of the model fit is inappropriate.
   */
  def fitModel(
      order:(Int, Int, Int),
      ts: Vector[Double],
      includeIntercept: Boolean = true,
      method: String = "css"): ARIMAModel = {
    val (p, d, q) = order
    val diffedTs = differences(ts, d).toArray.drop(d)

    if (p > 0 && q == 0) {
      val arModel = Autoregression.fitModel(new DenseVector(diffedTs), p)
      return new ARIMAModel(order, Array(arModel.c) ++ arModel.coefficients, includeIntercept)
    }

    val initParams = HannanRisannenInit(diffedTs, order, includeIntercept)
    val initCoeffs = fitWithCSS(diffedTs, order, includeIntercept, initParams)

    method match {
      case "css" => new ARIMAModel(order,initCoeffs, includeIntercept)
      case "ml" => throw new UnsupportedOperationException()
    }
  }

  /**
   * Fit an ARIMA model using conditional sum of squares, currently optimized using unbounded
   * BOBYQA.
   * @param diffedY time series we wish to fit to, already differenced if necessary
   * @param includeIntercept does the model include an intercept
   * @param initParams initial parameter guesses
   * @return
   */
  private def fitWithCSS(
      diffedY: Array[Double],
      order: (Int, Int, Int),
      includeIntercept: Boolean,
      initParams: Array[Double]): Array[Double]= {
    // We set up starting/ending trust radius using default suggested in
    // http://cran.r-project.org/web/packages/minqa/minqa.pdf
    // While # of interpolation points as mentioned common in
    // Source: http://www.damtp.cam.ac.uk/user/na/NA_papers/NA2009_06.pdf
    val (p, d, q) = order
    val radiusStart = math.min(0.96, 0.2 * initParams.map(math.abs).max)
    val radiusEnd = radiusStart * 1e-6
    val dimension = p + q + (if (includeIntercept) 1 else 0)
    val interpPoints = dimension * 2 + 1

    val optimizer = new BOBYQAOptimizer(interpPoints, radiusStart, radiusEnd)
    val objFunction = new ObjectiveFunction(new MultivariateFunction() {
      def value(params: Array[Double]): Double = {
        new ARIMAModel(order, params, includeIntercept).logLikelihoodCSSARMA(diffedY)
      }
    })

    val initialGuess = new InitialGuess(initParams)
    val maxIter = new MaxIter(10000)
    val maxEval = new MaxEval(10000)
    // TODO: Enforce stationarity and invertibility for AR and MA terms
    val bounds = SimpleBounds.unbounded(dimension)
    val goal = GoalType.MAXIMIZE
    val optimal = optimizer.optimize(objFunction, goal, bounds, maxIter, maxEval, initialGuess)
    optimal.getPoint
  }

  // TODO: implement MLE parameter estimates with Kalman filter.
  private def fitWithML(
      diffedY: Array[Double],
      arTerms: Array[Array[Double]],
      maTerms: Array[Double],
      initCoeffs: Array[Double]): Array[Double] = {
    throw new UnsupportedOperationException()
  }

  /**
   * Initializes ARMA parameter estimates using the Hannan-Risannen algorithm. The process is:
   * fit an AR(m) model of a higher order (i.e. m > max(p, q)), use this to estimate errors,
   * then fit an OLS model of AR(p) terms and MA(q) terms. The coefficients estimated by
   * the OLS model are returned as initial parameter estimates
   * @param diffedY differenced time series, as appropriate
   * @param order model order (p, d, q)
   * @param includeIntercept boolean indicating if an intercept should be fit as well
   * @return initial ARMA(p, d, q) parameter estimates
   */
  private def HannanRisannenInit(
      diffedY: Array[Double],
      order:(Int, Int, Int),
      includeIntercept: Boolean): Array[Double] = {
    val (p, d, q) = order
    val addToLag = 1
    val m = math.max(p, q) + addToLag // m > max(p, q)
    // higher order AR(m) model
    val arModel = Autoregression.fitModel(new DenseVector(diffedY), m)
    val arTerms1 = Lag.lagMatTrimBoth(diffedY, m, includeOriginal = false)
    val yTrunc = diffedY.drop(m)
    val estimated = arTerms1.zip(
      Array.fill(yTrunc.length)(arModel.coefficients)
      ).map { case (v, b) => v.zip(b).map { case (yi, bi) => yi * bi}.sum + arModel.c }
    // errors estimated from AR(m)
    val errors = yTrunc.zip(estimated).map { case (y, yhat) => y - yhat }
    // secondary regression, regresses X_t on AR and MA terms
    val arTerms2 = Lag.lagMatTrimBoth(yTrunc, p, includeOriginal = false).drop(math.max(q - p, 0))
    val errorTerms = Lag.lagMatTrimBoth(errors, q, includeOriginal = false).drop(math.max(p - q, 0))
    val allTerms = arTerms2.zip(errorTerms).map { case (ar, ma) => ar ++ ma }
    val regression = new OLSMultipleLinearRegression()
    regression.setNoIntercept(!includeIntercept)
    regression.newSampleData(yTrunc.drop(m - addToLag), allTerms)
    val params = regression.estimateRegressionParameters()
    params
  }

  /**
   * Calculate a differenced vector of a given order. Size-preserving by leaving first `order`
   * elements intact
   * @param ts Series to difference
   * @param order The difference order (e.g. x means y(i) = ts(i) - ts(i - x), etc)
   * @return a new differenced vector
   */
  def differences(ts: Vector[Double], order: Int): Vector[Double] = {
    if (order == 0) {
      // for consistency, since we create a new vector in else-branch
      ts.copy
    } else {
      val n = ts.length
      val diffedTs = new DenseVector(Array.fill(n)(0.0))
      var i = 0

      while (i < n) {
        // elements prior to `order` are copied over without modification
        diffedTs(i) = if (i < order) ts(i) else ts(i) - ts(i - order)
        i += 1
      }
      diffedTs
    }
  }

  /**
   * Calculate a "inverse-differenced" vector of a given order. Size-preserving by leaving first
   * `order` elements intact
   * @param ts Series to add up
   * @param order The difference order to add (e.g. x means y(i) = ts(i) + y(i -
   *              x), etc)
   * @return a new vector where the difference operation as been inverted
   */
  def invDifferences(ts: Vector[Double], order: Int): Vector[Double] = {

    if (order == 0) {
      // for consistency, since we create a new vector in else-branch
      ts.copy
    } else {
      val n = ts.length
      val addedTs = new DenseVector(Array.fill(n)(0.0))
      var i = 0

      while (i < n) {
        // elements prior to `order` are copied over without modification
        addedTs(i) = if (i < order) ts(i) else ts(i) + addedTs(i - order)
        i += 1
      }
      addedTs
    }
  }
}

class ARIMAModel(
    val order:(Int, Int, Int), // AR(p), differencing(d), MA(q)
    val coefficients: Array[Double], // coefficients: intercept, AR, MA, with increasing degrees
    val hasIntercept: Boolean = true) extends TimeSeriesModel {
  /**
   * loglikelihood based on conditional sum of squares
   * Source: http://www.nuffield.ox.ac.uk/economics/papers/1997/w6/ma.pdf
   * @param y time series
   * @return loglikehood
   */
  def logLikelihoodCSS(y: Vector[Double]): Double = {
    val d = order._2
    val diffedY = ARIMA.differences(y, d).toArray.drop(d)
    logLikelihoodCSSARMA(diffedY)
  }

  /**
   * loglikelihood based on conditional sum of squares. In contrast to
   * [[com.cloudera.sparkts.ARIMAModel.logLikelihoodCSS]] the array provided should correspond
   * to an already differenced array, so that the function below corresponds to the loglikelihood
   * for the ARMA rather than the ARIMA process
   * @param diffedY differenced array
   * @return loglikelihood of ARMA
   */
  def logLikelihoodCSSARMA(diffedY: Array[Double]): Double = {
    val n = diffedY.length
    val yHat = new DenseVector(Array.fill(n)(0.0))
    val yVec = new DenseVector(diffedY)
    iterateARMA(yVec,  yHat, _ + _, goldStandard = yVec)

    val (p, d, q) = order
    val maxLag = math.max(p, q)
    // drop first maxLag terms, since we can't estimate residuals there, since no
    // AR(n) terms available
    val css = diffedY.zip(yHat.toArray).drop(maxLag).map { case (obs, pred) =>
      math.pow(obs - pred, 2)
    }.sum
    val sigma2 = css / n
    (-n / 2) * math.log(2 * math.Pi * sigma2) - css / (2 * sigma2)
  }

  /**
   * Updates the error vector in place with a new (more recent) error
   * The newest error is placed in position 0, while older errors "fall off the end"
   * @param errs array of errors of length q in ARIMA(p, d, q), holds errors for t-1 through t-q
   * @param newError the error at time t
   * @return a modified array with the latest error placed into index 0
   */
  def updateMAErrors(errs: Array[Double], newError: Double): Unit= {
    val n = errs.length
    var i = 0
    while (i < n - 1) {
      errs(i + 1) = errs(i)
      i += 1
    }
    if (n > 0) {
      errs(0) = newError
    }
  }

  /**
   * Perform operations with the AR and MA terms, based on the time series `ts` and the errors
   * based off of `goldStandard` or `errors`, combined with elements from the series  `dest`.
   * Weights for terms are taken from the current model configuration.
   * So for example: iterateARMA(series1, series_of_zeros,  _ + _ , goldStandard = series1,
   * initErrors = null)
   * calculates the 1-step ahead forecasts for series1 assuming current coefficients, and initial
   * MA errors of 0.
   * @param ts Time series to use for AR terms
   * @param dest Time series holding initial values at each index
   * @param op Operation to perform between values in dest, and various combinations of ts, errors
   *           and intercept terms
   * @param goldStandard The time series to which to compare 1-step ahead forecasts to obtain
   *                     moving average errors. Default set to null. In which case, we check if
   *                     errors are directly provided. Either goldStandard or errors can be null,
   *                     but not both
   * @param errors The time series of errors to be used as error at time i, which is then used
   *               for moving average terms in future indices. Either goldStandard or errors can
   *               be null, but not both
   * @param initMATerms Initialization for first q terms of moving average. If none provided (i.e.
   *                    remains null, as per default), then initializes to all zeros
   * @return the time series resulting from the interaction of the parameters with the model's
   *         coefficients.
   */
  private def iterateARMA(
      ts: Vector[Double],
      dest: Vector[Double],
      op: (Double, Double) => Double,
      goldStandard: Vector[Double] = null,
      errors: Vector[Double] = null,
      initMATerms: Array[Double] = null): Vector[Double] = {
    require(goldStandard != null || errors != null, "goldStandard or errors must be passed in")
    val (p, d, q) = order
    val maTerms = if (initMATerms == null) Array.fill(q)(0.0) else initMATerms
    val intercept = if (hasIntercept) 1 else 0
    var i = math.max(p, q) // maximum lag
    var j = 0
    val n = ts.length
    var error = 0.0

    while (i < n) {
      j = 0
      // intercept
      dest(i) = op(dest(i), intercept * coefficients(j))
      // autoregressive terms
      while (j < p && i - j - 1 >= 0) {
        dest(i) = op(dest(i), ts(i - j - 1) * coefficients(intercept + j))
        j += 1
      }
      // moving average terms
      j = 0
      while (j < q) {
        dest(i) = op(dest(i), maTerms(j) * coefficients(intercept + p + j))
        j += 1
      }

      error = if (goldStandard == null) errors(i) else goldStandard(i) - dest(i)
      updateMAErrors(maTerms, error)
      i += 1
    }
    dest
  }

  /**
   * Given a timeseries, assume that it is the result of an ARIMA(p, d, q) process, and apply
   * inverse operations to obtain the original series of underlying errors.
   * To do so, we assume prior MA terms are 0.0, and prior AR are equal to the model's intercept
   * @param ts Time series of observations with this model's characteristics.
   * @param destTs
   * @return The dest series, representing remaining errors, for convenience.
   */
  def removeTimeDependentEffects(ts: Vector[Double], destTs: Vector[Double]): Vector[Double] = {
    val (p, d, q) = order
    // difference as necessary
    val diffed = ARIMA.differences(ts, d)
    val maxLag = math.max(p, q)
    val interceptAmt = if (hasIntercept) coefficients(0) else 0.0
    // extend vector so that initial AR(maxLag) are equal to intercept value
    val diffedExtended = new DenseVector(Array.fill(maxLag)(interceptAmt) ++ diffed.toArray)
    val changes = diffedExtended.copy
    // changes(i) corresponds to the error term at index i, since when it is used we will have
    // removed AR and MA terms at index i
    iterateARMA(diffedExtended, changes, _ - _, errors = changes)
    destTs := changes(maxLag to -1)
    destTs
  }

  /**
   * Given a timeseries, apply an ARIMA(p, d, q) model to it.
   * We assume that prior MA terms are 0.0 and prior AR terms are equal to the intercept
   * @param ts Time series of i.i.d. observations.
   * @param destTs
   * @return The dest series, representing the application of the model to provided error
   *         terms, for convenience.
   */
  def addTimeDependentEffects(ts: Vector[Double], destTs: Vector[Double]): Vector[Double] = {
    val (p, d, q) = order
    val maxLag = math.max(p, q)
    val interceptAmt = if (hasIntercept) coefficients(0) else 0.0
    // extend vector so that initial AR(maxLag) are equal to intercept value
    // Recall iterateARMA begins at index maxLag, and uses previous values as AR terms
    val changes = new DenseVector(Array.fill(maxLag)(interceptAmt) ++ ts.toArray)
    // ts corresponded to errors, and we want to use them
    val errorsProvided = changes.copy
    iterateARMA(changes, changes, _ + _, errors = errorsProvided)
    // perform any inverse differencing required
    destTs := ARIMA.invDifferences(changes(maxLag to -1), d)
    destTs
  }

  /**
   * Sample a series of size n assuming an ARIMA(p, d, q) process.
   * @param n size of sample
   * @return series reflecting ARIMA(p, d, q) process
   */
  def sample(n: Int, rand: RandomGenerator): Vector[Double] = {
    val vec = new DenseVector(Array.fill[Double](n)(rand.nextGaussian))
    addTimeDependentEffects(vec, vec)
  }

  /**
   * Provided fitted values for timeseries ts as 1-step ahead forecasts, based on current
   * model parameters, and then provide `nFuture` periods of forecast. We assume AR terms
   * prior to the start of the series are equal to the model's intercept term. Meanwhile, MA
   * terms prior to the start are assumed to be 0.0
   * @param ts Timeseries to use as gold-standard. Each value (i) in the returning series
   *           is a 1-step ahead forecast of ts(i). We use the difference between ts(i) -
   *           estimated(i) to calculate the error at time i, which is used for the moving
   *           average terms
   * @param nFuture Periods in the future to forecast (beyond length of ts)
   * @return a series consisting of fitted 1-step ahead forecasts for historicals and then
   *         `nFuture` periods of forecasts. Note that in the future values error terms become
   *         zero and prior predictions are used for any AR terms.
   */
  def forecast(ts: Vector[Double], nFuture: Int): Vector[Double] = {
    val (p, d, q) = order
    val maxLag = math.max(p, q)
    // difference timeseries as necessary for model
    val diffedTs = new DenseVector(ARIMA.differences(ts, d).toArray.drop(d))

    // Assumes prior AR terms are equal to model intercept
    val interceptAmt =  if (hasIntercept) coefficients(0) else 0.0
    val diffedTsExtended = new DenseVector(Array.fill(maxLag)(interceptAmt) ++ diffedTs.toArray)
    val histLen = diffedTsExtended.length
    val hist = new DenseVector(Array.fill(histLen)(0.0))

    // fit historical values
    iterateARMA(diffedTsExtended, hist, _ + _,  goldStandard = diffedTsExtended)

    // Last set of errors, to be used in forecast if MA terms included
    val maTerms = (for (i <- histLen - maxLag until histLen) yield {
      diffedTsExtended(i) - hist(i)
    }).toArray

    // copy over last maxLag values, to use in iterateARMA for forward curve
    val forward = new DenseVector(Array.fill(nFuture + maxLag)(0.0))
    forward(0 until maxLag) := hist(-maxLag to -1)
    // use self as ts to take AR from same series, use self as goldStandard to induce future errors
    // of zero, and use prior moving average errors as initial error terms for MA
    iterateARMA(forward, forward, _ + _, goldStandard = forward, initMATerms = maTerms)

    val results = new DenseVector(Array.fill(ts.length + nFuture)(0.0))
    // drop first maxLag terms from historicals, since these are our assumed AR terms
    results(0 until ts.length) := hist(maxLag to -1)
    // drop first maxLag terms from forward curve, these are part of hist already
    results(ts.length to -1) :=  forward(maxLag to -1)
    // add up if there was any differencing required
    ARIMA.invDifferences(results, d)
  }
}

File: python/sparkts/models/ARIMA.py
from . import _py2java_double_array
from _model import PyModel

from pyspark.mllib.common import _py2java, _java2py
from pyspark.mllib.linalg import Vectors


"""
ARIMA models allow modeling timeseries as a function of prior values of the series
(i.e. autoregressive terms) and a moving average of prior error terms. ARIMA models
are traditionally specified as ARIMA(p, d, q), where p is the autoregressive order,
d is the differencing order, and q is the moving average order. Using the backshift (aka
lag operator) B, which when applied to a Y returns the prior value, the
ARIMA model can be specified as
Y_t = c + \sum_{i=1}^p \phi_i*B^i*Y_t + \sum_{i=1}^q \theta_i*B^i*\epsilon_t + \epsilon_t
where Y_i has been differenced as appropriate according to order `d`
See [[https://en.wikipedia.org/wiki/Autoregressive_integrated_moving_average]] for more
information on ARIMA.
See [[https://en.wikipedia.org/wiki/Order_of_integration]] for more information on differencing
integrated time series.


"""

def autofit(ts, maxp=5, maxd=2, maxq=5, sc=None):
    """
    Utility function to help in fitting an automatically selected ARIMA model based on approximate
    Akaike Information Criterion (AIC) values. The model search is based on the heuristic
    developed by Hyndman and Khandakar (2008) and described in [[http://www.jstatsoft
    .org/v27/i03/paper]]. In contrast to the algorithm in the paper, we use an approximation to
    the AIC, rather than an exact value. Note that if the maximum differencing order provided
    does not suffice to induce stationarity, the function returns a failure, with the appropriate
    message. Additionally, note that the heuristic only considers models that have parameters
    satisfying the stationarity/invertibility constraints. Finally, note that our algorithm is
    slightly more lenient than the original heuristic. For example, the original heuristic
    rejects models with parameters "close" to violating stationarity/invertibility. We only
    reject those that actually violate it.
   
    This functionality is even less mature than some of the other model fitting functions here, so
    use it with caution.
   
    Parameters
    ----------
    ts:
        time series to which to automatically fit an ARIMA model as a Numpy array
    maxP:
        limit for the AR order
    maxD:
        limit for differencing order
    maxQ:
        limit for the MA order
    sc:
        The SparkContext, required.
    
    returns an ARIMAModel
    """
    assert sc != None, "Missing SparkContext"
    
    jmodel = sc._jvm.com.cloudera.sparkts.models.ARIMA.autoFit(_py2java(sc, Vectors.dense(ts)), maxp, maxd, maxq)
    return ARIMAModel(jmodel=jmodel, sc=sc)

def fit_model(p, d, q, ts, includeIntercept=True, method="css-cgd", userInitParams=None, sc=None):
    """
    Given a time series, fit a non-seasonal ARIMA model of order (p, d, q), where p represents
    the autoregression terms, d represents the order of differencing, and q moving average error
    terms. If includeIntercept is true, the model is fitted with an intercept. In order to select
    the appropriate order of the model, users are advised to inspect ACF and PACF plots, or compare
    the values of the objective function. Finally, while the current implementation of
    `fitModel` verifies that parameters fit stationarity and invertibility requirements,
    there is currently no function to transform them if they do not. It is up to the user
    to make these changes as appropriate (or select a different model specification)

    Parameters
    ----------
    p:
        autoregressive order
    d:
        differencing order
    q:
        moving average order
    ts:
        time series to which to fit an ARIMA(p, d, q) model as a Numpy array.
    includeIntercept:
        if true the model is fit with an intercept term. Default is true
    method:
        objective function and optimization method, current options are 'css-bobyqa',
        and 'css-cgd'. Both optimize the log likelihood in terms of the
        conditional sum of squares. The first uses BOBYQA for optimization, while
        the second uses conjugate gradient descent. Default is 'css-cgd'
    userInitParams:
        A set of user provided initial parameters for optimization as a float list.
        If null (default), initialized using Hannan-Rissanen algorithm. If provided,
        order of parameter should be: intercept term, AR parameters (in
        increasing order of lag), MA parameters (in increasing order of lag).
    sc:
        The SparkContext, required.
    
    returns an ARIMAModel
    """
    assert sc != None, "Missing SparkContext"
    
    jvm = sc._jvm
    jmodel = jvm.com.cloudera.sparkts.models.ARIMA.fitModel(p, d, q, _py2java(sc, Vectors.dense(ts)), includeIntercept, method, _py2java_double_array(sc, userInitParams))
    return ARIMAModel(jmodel=jmodel, sc=sc)

class ARIMAModel(PyModel):
    def __init__(self, p=0, d=0, q=0, coefficients=None, hasIntercept=False, jmodel=None, sc=None):
        assert sc != None, "Missing SparkContext"

        self._ctx = sc
        if jmodel == None:
            self._jmodel = self._ctx._jvm.com.cloudera.sparkts.models.ARIMAModel(p, d, q, _py2java_double_array(self._ctx, coefficients), hasIntercept)
        else:
            self._jmodel = jmodel
            
        self.p = _java2py(sc, self._jmodel.p())
        self.d = _java2py(sc, self._jmodel.d())
        self.q = _java2py(sc, self._jmodel.q())
        self.coefficients = _java2py(sc, self._jmodel.coefficients())
        self.has_intercept = _java2py(sc, self._jmodel.hasIntercept())
    
    def log_likelihood_css(self, y):
        """
        log likelihood based on conditional sum of squares
        
        Source: http://www.nuffield.ox.ac.uk/economics/papers/1997/w6/ma.pdf

        Parameters
        ----------
        y:
            time series as a DenseVector

        returns log likelihood as a double
        """
        likelihood = self._jmodel.logLikelihoodCSS(_py2java(self._ctx, y))
        return _java2py(self._ctx, likelihood)

    def log_likelihood_css_arma(self, diffedy):
        """
        log likelihood based on conditional sum of squares. In contrast to logLikelihoodCSS the array
        provided should correspond to an already differenced array, so that the function below
        corresponds to the log likelihood for the ARMA rather than the ARIMA process
        
        Parameters
        ----------
        diffedY:
            differenced array as a numpy array
        
        returns log likelihood of ARMA as a double
        """
        # need to copy diffedy to a double[] for Java
        likelihood =  self._jmodel.logLikelihoodCSSARMA(_py2java_double_array(self._ctx, diffedy))
        return _java2py(self._ctx, likelihood)
        
    def gradient_log_likelihood_css_arma(self, diffedy):
        """
        Calculates the gradient for the log likelihood function using CSS
        Derivation:
            L(y | \theta) = -\frac{n}{2}log(2\pi\sigma^2) - \frac{1}{2\pi}\sum_{i=1}^n \epsilon_t^2 \\
            \sigma^2 = \frac{\sum_{i = 1}^n \epsilon_t^2}{n} \\
            \frac{\partial L}{\partial \theta} = -\frac{1}{\sigma^2}
            \sum_{i = 1}^n \epsilon_t \frac{\partial \epsilon_t}{\partial \theta} \\
            \frac{\partial \epsilon_t}{\partial \theta} = -\frac{\partial \hat{y}}{\partial \theta} \\
            \frac{\partial\hat{y}}{\partial c} = 1 +
            \phi_{t-q}^{t-1}*\frac{\partial \epsilon_{t-q}^{t-1}}{\partial c} \\
            \frac{\partial\hat{y}}{\partial \theta_{ar_i}} =  y_{t - i} +
            \phi_{t-q}^{t-1}*\frac{\partial \epsilon_{t-q}^{t-1}}{\partial \theta_{ar_i}} \\
            \frac{\partial\hat{y}}{\partial \theta_{ma_i}} =  \epsilon_{t - i} +
            \phi_{t-q}^{t-1}*\frac{\partial \epsilon_{t-q}^{t-1}}{\partial \theta_{ma_i}} \\
        
        Parameters
        ----------
        diffedY:
            array of differenced values
        
        returns the gradient log likelihood as an array of double
        """
        # need to copy diffedy to a double[] for Java
        result =  self._jmodel.gradientlogLikelihoodCSSARMA(_py2java_double_array(self._ctx, diffedy))
        return _java2py(self._ctx, result)
    
    def sample(self, n):
        """
        Sample a series of size n assuming an ARIMA(p, d, q) process.
        
        Parameters
        ----------
        n:
            size of sample
            
        Returns a series reflecting ARIMA(p, d, q) process as a DenseVector
        """
        rg = self._ctx._jvm.org.apache.commons.math3.random.JDKRandomGenerator()
        return _java2py(self._ctx, self._jmodel.sample(n, rg))
    
    def forecast(self, ts, nfuture):
        """
        Provided fitted values for timeseries ts as 1-step ahead forecasts, based on current
        model parameters, and then provide `nFuture` periods of forecast. We assume AR terms
        prior to the start of the series are equal to the model's intercept term (or 0.0, if fit
        without and intercept term).Meanwhile, MA terms prior to the start are assumed to be 0.0. If
        there is differencing, the first d terms come from the original series.
       
        Parameters
        ----------
        ts:
            Timeseries to use as gold-standard. Each value (i) in the returning series
            is a 1-step ahead forecast of ts(i). We use the difference between ts(i) -
            estimate(i) to calculate the error at time i, which is used for the moving
            average terms.
        nFuture:
            Periods in the future to forecast (beyond length of ts)
            
        Returns a series consisting of fitted 1-step ahead forecasts for historicals and then
        `nFuture` periods of forecasts. Note that in the future values error terms become
        zero and prior predictions are used for any AR terms.
        
        """
        jts = _py2java(self._ctx, ts)
        jfore = self._jmodel.forecast(jts, nfuture)
        return _java2py(self._ctx, jfore)
    
    def is_stationary(self):
        """
        Check if the AR parameters result in a stationary model. This is done by obtaining the roots
        to the polynomial 1 - phi_1 * x - phi_2 * x^2 - ... - phi_p * x^p = 0, where phi_i is the
        corresponding AR parameter. Note that we check the roots, not the inverse of the roots, so
        we check that they lie outside of the unit circle.
        Always returns true for models with no AR terms.
        See http://www.econ.ku.dk/metrics/Econometrics2_05_II/Slides/07_univariatetimeseries_2pp.pdf
        for more information (specifically slides 23 - 25)
        
        Returns true if the model's AR parameters are stationary
        """
        return self._jmodel.isStationary()
    
    def is_invertible(self):
        """
        Checks if MA parameters result in an invertible model. Checks this by solving the roots for
        1 + theta_1 * x + theta_2 * x + ... + theta_q * x&#94;q = 0. Please see
        [[ARIMAModel.isStationary]] for more details.
        Always returns true for models with no MA terms.
        
        Returns true if the model's MA parameters are invertible
        """
        return self._jmodel.isInvertible()
    
    def approx_aic(self, ts):
        """
        Calculates an approximation to the Akaike Information Criterion (AIC). This is an approximation
        as we use the conditional likelihood, rather than the exact likelihood. Please see
        [[https://en.wikipedia.org/wiki/Akaike_information_criterion]] for more information on this
        measure.
        
        Parameters
        ----------
        ts:
            the timeseries to evaluate under current model
            
        Returns an approximation to the AIC under the current model as a double
        """
        return self._jmodel.approxAIC(_py2java(self._ctx, ts))
